metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowExecutionDao.java

```diff
@@ -346,6 +354,122 @@ private AbstractMetisPlugin getFirstOrLastFinishedWorkflowExecutionPluginByDatas
+      int nextPage) {
+
+    // Create the aggregate pipeline
+    final AggregationPipeline pipeline = morphiaDatastoreProvider.getDatastore()
+        .createAggregation(WorkflowExecution.class);
+
+    // Step 1: match the datasets that are in the list.
+    if (datasetIds != null) {
+      final Query<WorkflowExecution> query =
+          morphiaDatastoreProvider.getDatastore().createQuery(WorkflowExecution.class);
+      query.field(DATASET_ID).in(datasetIds);
+      pipeline.match(query);
+    }
+
+    // Step 2: Add difference integers on whether the status is INQUEUE or RUNNING.
+    final String statusInQueueField = "statusInQueue";
+    final String statusRunningField = "statusRunning";
+    final int inqueueDifferennce = DEFAULT_POSITION_IN_OVERVIEW - INQUEUE_POSITION_IN_OVERVIEW;
+    final int runningDifferennce = DEFAULT_POSITION_IN_OVERVIEW - RUNNING_POSITION_IN_OVERVIEW;
+    pipeline.project(
+        Projection.projection(statusInQueueField, Projection.expression("$cond",
+            Projection.expression(FilterOperator.EQUAL.val(), "INQUEUE", "$" + WORKFLOW_STATUS),
+            inqueueDifferennce, 0)),
+        Projection.projection(statusRunningField, Projection.expression("$cond",
+            Projection.expression(FilterOperator.EQUAL.val(), "RUNNING", "$" + WORKFLOW_STATUS),
+            runningDifferennce, 0)),
+        Projection.projection(OrderField.CREATED_DATE.getOrderFieldName()),
+        Projection.projection(DATASET_ID)
+    );
+
+    // Step 3: Subtract these integers from the default value to obtain the right ordering.
+    final String statusIndexField = "statusIndex";
+    pipeline.project(
+        Projection.projection(statusIndexField, Projection.subtract(DEFAULT_POSITION_IN_OVERVIEW,
+            Projection.add("$" + statusInQueueField, "$" + statusRunningField))),
+        Projection.projection(OrderField.CREATED_DATE.getOrderFieldName()),
+        Projection.projection(DATASET_ID)
+    );
+
+    // Step 4: Sort - first on the status index, then on the createdDate.
+    pipeline.sort(Sort.ascending(statusIndexField),
+        Sort.descending(OrderField.CREATED_DATE.getOrderFieldName()));
+
+    // Step 5: Apply pagination
+    pipeline.skip(nextPage * getWorkflowExecutionsPerRequest())
+        .limit(getWorkflowExecutionsPerRequest());
+
+    // Step 6: Join with the dataset and the execution
+    final String datasetCollectionName = morphiaDatastoreProvider.getDatastore()
+        .getCollection(Dataset.class).getName();
+    final String executionCollectionName = morphiaDatastoreProvider.getDatastore()
+        .getCollection(WorkflowExecution.class).getName();
+    final String datasetListField = "datasetList";
+    final String executionListField = "executionList";
+    pipeline.lookup(datasetCollectionName, DATASET_ID, DatasetDao.DATASET_ID, datasetListField);
+    pipeline.lookup(executionCollectionName, "_id", "_id", executionListField);
```

Comment:
Why is this join required?
Is it that the first lookup(Join) will produce the workflowExecution with the addition of the "datasetList" which in fact can only contain one item, the one that comes from the dataset collection?

Reply:
As discussed: we don't have the execution here: by repeatedly projecting we lose all the fields we don't explicitly name. This is better than naming all those fields.

Reply:
Yes